File: EconSim/src/io/ReadLists.kt
package io

import io.Read.Companion.MalformedItemException
import io.Read.Companion.NoSuchTypeException
import io.Read.Companion.UnexpectedCharacterException

inline fun <reified T> readList(name: String, type: String, tokenIterator: Iterator<String>): List<T> {
    val listOfAny = mutableListOf<Any>()
    while (tokenIterator.hasNext()) {
        val nextListElement: Any? = readNextListElement(name, type, tokenIterator)
        if (nextListElement == null) {
            return filterList(name, type, listOfAny)
        } else {
            listOfAny.add(nextListElement)
            if (tokenIterator.hasNext().not()) break
            when (tokenIterator.next()) {
                "]" -> return filterList(name, type, listOfAny)
                "," -> {}
                else -> throw MalformedItemException("list $name", "must use commas to separate list elements")
            }
        }
    }
    throw MalformedItemException("list $name", "has no closing ']'")
}

fun readNextListElement(listName: String, type: String, tokenIterator: Iterator<String>): Any? {
    if (tokenIterator.hasNext().not()) throw MalformedItemException("list $listName", "has no closing ']'")
    when (val token = tokenIterator.next()) {
        "=" -> throw UnexpectedCharacterException('=', "list $listName", "cannot use assignment in list")
        "}" -> throw UnexpectedCharacterException('}', "list $listName", "cannot close undeclared object in list")
        "[" -> throw UnexpectedCharacterException('[', "list $listName", "cannot declare list in list")
        "(" -> throw UnexpectedCharacterException('(', "list $listName", "cannot declare primitive in list")
        ")" -> throw UnexpectedCharacterException(')', "list $listName", "cannot declare primitive in list")
        "," -> throw UnexpectedCharacterException(',', "list $listName", "cannot have empty element in list")
        "{" -> {return 1}
        "]" -> return null
        else -> {
            return when (type) {
                "Int" -> readInt(token)
                "Float" -> readFloat(token)
                "String" -> token
                else -> throw NoSuchTypeException(type)
            }
        }
    }
}

inline fun <reified T> filterList(listName: String, type: String, listOfAny: List<Any>): List<T> {
    val listOfT = listOfAny.filterIsInstance<T>()
    if (listOfT.size == listOfAny.size) return listOfT
    else throw MalformedItemException("list $listName", "cannot contain elements not of type $type")
}


inline fun <reified T> readList2(name: String, type: String, tokenIterator: Iterator<String>): List<T> {
    val listOfAny = handleList2(name, type, tokenIterator)
    val listOfT = mutableListOf<T>()

    for (element in listOfAny) {
        if (element !is T) throw MalformedItemException("list $name", "cannot contain elements not of type $type")
        listOfT.add(element)
    }
    return listOfT.toList()
}

fun handleList2(name: String, type: String, tokenIterator: Iterator<String>): List<Any> {
    val listOfElements = mutableListOf<Any>()
    var currentElement: Any? = null
    while (tokenIterator.hasNext()) {
        when (val token = tokenIterator.next()) {
            "=" -> throw UnexpectedCharacterException('=', "list $name", "cannot use assignment in list")
            "}" -> throw UnexpectedCharacterException('}', "list $name", "cannot close undeclared object in list")
            "[" -> throw UnexpectedCharacterException('[', "list $name", "cannot declare list in list")
            "(" -> throw UnexpectedCharacterException('(', "list $name", "cannot declare primitive in list")
            ")" -> throw UnexpectedCharacterException(')', "list $name", "cannot declare primitive in list")

            "{" -> {
                if (currentElement != null) throw UnexpectedCharacterException('{', "list $name", "list elements must be separated with commas")
            }

            "," -> {
                if (currentElement == null) throw UnexpectedCharacterException(',', "list $name", "cannot have empty element in list")
                listOfElements.add(currentElement)
                currentElement = null
            }

            "]" -> {
                if (currentElement != null) listOfElements.add(currentElement)
                return listOfElements.toList()
            }

            else -> {
                if (currentElement != null) throw MalformedItemException("list $name", "must separate list items with commas")
                currentElement = handleListPrimitive2(type, token)
            }
        }
    }
    throw MalformedItemException("list $name", "has no closing ']'")
}

private fun handleListPrimitive2(type: String, token: String): Any {
    when (type) {
        "Int" -> return readInt(token)
        "Float" -> return readFloat(token)
        "String" -> return token
    }
    throw NoSuchTypeException(type)
}

File: Populo/Assets/Scripts/IO/Read/ReadLists.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

public class ReadLists
{
    public static List<Color32> ReadColorList(string name, FileTokenIterator tokenIterator)
    {
        return new List<Color32>(
            from iterator
            in ReadListElements(name, new Color32().GetTypeName(), tokenIterator)
            select ReadPrimitives.ReadColor(iterator)
        );
    }

    public static List<float> ReadFloatList(string name, FileTokenIterator tokenIterator)
    {
        return new List<float>(
            from iterator
            in ReadListElements(name, new float().GetTypeName(), tokenIterator)
            select ReadPrimitives.ReadFloat(iterator)
        );
    }

    public static List<int> ReadIntList(string name, FileTokenIterator tokenIterator)
    {
        return new List<int>(
            from iterator
            in ReadListElements(name, new int().GetTypeName(), tokenIterator)
            select ReadPrimitives.ReadInt(iterator)
        );
    }

    public static List<string> ReadStringList(string name, FileTokenIterator tokenIterator)
    {
        return new List<string>(
            from iterator
            in ReadListElements(name, "".GetTypeName(), tokenIterator)
            select ReadPrimitives.ReadString(iterator)
        );
    }

    private static FileTokenIterator[] ReadListElements(string listName, string listType, FileTokenIterator tokenIterator)
    {
        List<FileTokenIterator> listElements = new List<FileTokenIterator>();
        while (tokenIterator.HasNext())
        {
            FileTokenIterator nextListElement = ReadListElement(listName, listType, tokenIterator);
            if (nextListElement == null) break;
            listElements.Add(nextListElement);
            if (IsEndOfList()) break;
        }
        return listElements.ToArray();

        bool IsEndOfList()
        {
            if (!tokenIterator.HasNext()) return true;
            if (tokenIterator.Next() == ",") return false;
            else throw new FileTokenIterator.MalformedItemException($"list {listName}", "must have commas to separate list elements", tokenIterator);
        }
    }

    private static FileTokenIterator ReadListElement(string listName, string listType, FileTokenIterator tokenIterator)
    {
        if (!tokenIterator.HasNext()) throw new FileTokenIterator.MalformedItemException($"list {listName}", "has no closing ']'", tokenIterator);

        switch (tokenIterator.Next())
        {
            case "=":
                throw new FileTokenIterator.UnexpectedCharacterException('=', $"list {listName}", "cannot use assignment in list", tokenIterator);
            case "[":
                throw new FileTokenIterator.UnexpectedCharacterException('[', $"list {listName}", "cannot declare list in list", tokenIterator);
            case "}":
                throw new FileTokenIterator.UnexpectedCharacterException('}', $"list {listName}", "cannot close undeclared object in list", tokenIterator);
            case ")":
                throw new FileTokenIterator.UnexpectedCharacterException(')', $"list {listName}", "cannot declare primitive in list", tokenIterator);
            case ",":
                throw new FileTokenIterator.UnexpectedCharacterException(',', $"list {listName}", "cannot have empty element in list", tokenIterator);

            //case "{":
                //return FileTokenIterator.MakeFileTokenIteratorForCurrentObject(tokenIterator, $"{listName}", listType); ;
            case "(":
                return FileTokenIterator.MakeFileTokenIteratorForCurrentPrimitive(tokenIterator, $"{listName}", listType);

            default:
                throw new FileTokenIterator.MalformedItemException($"list {listName}", $"unrecognized item {tokenIterator.Current()}, must declare list elements as objects or primitives", tokenIterator);
        }
    }
}
